pkg/kubefed2/unjoin.go

```diff
@@ -163,76 +164,25 @@ func UnjoinCluster(hostConfig, clusterConfig *rest.Config, federationNamespace,
-		if err != nil {
-			return err
-		}
-	}
-
-	glog.V(2).Infof("Deleting federated cluster resource from namespace: %s for unjoin cluster: %s",
-		federationNamespace, unjoiningClusterName)
-
-	_, err = deleteFederatedClusterAndSecret(hostClientset, fedClientset, federationNamespace,
-		unjoiningClusterName, dryRun)
-	if err != nil {
-		glog.V(2).Infof("Failed to delete federated cluster resource from namespace: %s for unjoin cluster: %s due to: %v",
-			federationNamespace, unjoiningClusterName, err)
-		return err
+		removeFromClusterRegistry(hostConfig, clusterNamespace, clusterConfig.Host, unjoiningClusterName, dryRun)
 	}
 
-	glog.V(2).Infof("Deleted federated cluster resource from namespace: %s for unjoin cluster: %s",
-		federationNamespace, unjoiningClusterName)
-
-	err = deleteRBACResources(hostClientset, clusterClientset,
-		federationNamespace, unjoiningClusterName, hostClusterContext, dryRun)
-	if err != nil {
-		glog.V(2).Infof("Could not delete cluster RBAC resources: %v", err)
-		return err
-	}
+	deletionSucceeded := deleteRBACResources(clusterClientset, federationNamespace, unjoiningClusterName, hostClusterContext, dryRun)
 
 	if unjoiningClusterContext == hostClusterContext {
 		glog.V(2).Infof("Host cluster: %s and unjoin cluster: %s are same, no need to delete federation namespace: %s from unjoin cluster.",
 			hostClusterContext, unjoiningClusterContext, federationNamespace)
 	} else {
-		glog.V(2).Infof("Deleting federation namespace: %s from unjoin cluster: %s",
-			federationNamespace, unjoiningClusterName)
-
-		err = deleteFedNSFromUnjoinCluster(clusterClientset,
-			federationNamespace, unjoiningClusterName, hostClusterContext, dryRun)
+		err = deleteFedNSFromUnjoinCluster(clusterClientset, federationNamespace, unjoiningClusterName, dryRun)
 		if err != nil {
-			glog.V(2).Infof("Could not delete cluster RBAC resources: %v", err)
-			return err
+			deletionSucceeded = false
 		}
-
-		glog.V(2).Infof("Deleted federation namespace: %s from unjoin cluster: %s",
-			federationNamespace, unjoiningClusterName)
 	}
 
-	return nil
-}
-
-// performPreflightChecksForUnjoin checks that the host and unjoining clusters are in
-// a consistent state.
-func performPreflightChecksForUnjoin(clusterClientset client.Interface, name, hostClusterContext,
-	federationNamespace string) error {
-	// Make sure there is a existing service account in the unjoining cluster.
-	saName := util.ClusterServiceAccountName(name, hostClusterContext)
-	sa, err := clusterClientset.CoreV1().ServiceAccounts(federationNamespace).Get(saName,
-		metav1.GetOptions{})
-
-	if errors.IsNotFound(err) {
-		return fmt.Errorf("service account does not exist in unjoining cluster")
-	} else if err != nil {
-		return err
-	} else if sa != nil {
-		return nil
+	// deletionSucceeded when all operations in deleteRBACResources and deleteFedNSFromUnjoinCluster succeed.
```

Comment:
Please move this after L171, we should not delete namespace before delete resources in the namespace.

Reply:
The namespace in question is in the member cluster targeted for unjoining.  Why should it not be removed first?